File: app-server/src/test/kotlin/com/github/arhor/simple/expense/tracker/web/error/ErrorCodeSerializerTest.java
package com.github.arhor.simple.expense.tracker.web.error;

import java.io.IOException;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.SerializerProvider;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.BDDMockito.then;

@ExtendWith(MockitoExtension.class)
class ErrorCodeSerializerTest {

    private static ErrorCodeSerializer errorCodeSerializer;

    @Mock
    private JsonGenerator generator;

    @Mock
    private SerializerProvider provider;

    @Captor
    private ArgumentCaptor<String> stringCaptor;

    @BeforeAll
    static void setUpClass() {
        errorCodeSerializer = new ErrorCodeSerializer();
    }

    @AfterAll
    static void tearDownClass() {
        errorCodeSerializer = null;
    }

    @ParameterizedTest
    @EnumSource(ErrorCode.class)
    void each_error_code_should_be_correctly_serialized_to_string(final ErrorCode errorCode) throws IOException {
        // when
        errorCodeSerializer.serialize(errorCode, generator, provider);

        // then
        then(generator)
            .should()
            .writeString(stringCaptor.capture());

        then(provider)
            .shouldHaveNoInteractions();

        assertThat(stringCaptor.getValue())
            .isNotBlank()
            .contains(
                String.valueOf(errorCode.getType()),
                String.valueOf(errorCode.getNumericValue())
            );
    }
}

File: app-server/src/test/kotlin/com/github/arhor/simple/expense/tracker/web/error/ErrorCodeSerializerTest.kt
package com.github.arhor.simple.expense.tracker.web.error

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.SerializerProvider
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.runs
import io.mockk.slot
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.EnumSource

internal class ErrorCodeSerializerTest {

    private val errorCodeSerializer = ErrorCodeSerializer()

    private val generator = mockk<JsonGenerator>()
    private val provider = mockk<SerializerProvider>()
    private val serializedValue = slot<String>()

    @BeforeEach
    fun setUp() {
        serializedValue.clear()
        every { generator.writeString(any<String>()) } just runs
    }

    @ParameterizedTest
    @EnumSource(ErrorCode::class)
    fun `each error code should be correctly serialized to string`(
        // given
        errorCode: ErrorCode
    ) {
        // when
        errorCodeSerializer.serialize(errorCode, generator, provider)

        // then
        verify(exactly = 1) { generator.writeString(capture(serializedValue)) }

        assertThat(serializedValue.captured)
            .isNotBlank
            .contains(
                errorCode.type.toString(),
                errorCode.value.toString(),
            )
    }
}
